crates/common/tedge_utils/src/fs_notify.rs

```diff
@@ -63,57 +64,71 @@ pub enum NotifyStreamError {
 
     #[error("Watcher: {mask} is duplicated for file: {path:?}")]
     DuplicateWatcher { mask: FileEvent, path: PathBuf },
+
+    #[error("Wrong Event triggered")]
+    WrongEventTriggered,
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct EventDescription {
+    pub dir_path: PathBuf,
+    pub file_name: Option<String>,
+    pub masks: HashSet<FileEvent>,
 }
 
-#[derive(Debug, Default, Clone)]
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
 struct WatchDescriptor {
-    description: HashMap<PathBuf, HashMap<Option<String>, Vec<FileEvent>>>,
+    description: HashMap<c_int, Vec<EventDescription>>,
```

Comment:
Yes, you are right that `HashMap<c_int, EventDescription>` should be what we need here however sadly this is not possible because if a file does not exist you can not watch it. There is no `c_int` for this event.

Because of this issue, internally all watchers are directory watchers. 

This means that `c_int` can map to one or more events, hence the use of vector. When an inotify event is triggered, this event is matched on two things:
1. the `c_int`
2. the [name](https://docs.rs/inotify/latest/inotify/struct.Event.html#structfield.name) of the file the event was triggered for (which is an Option<file_name>)

Reply:
Yes, there will be one `event id` for watching different files in the same `directory`.
For example in a directory `/etc/tedge/operations/c8y` if we want to watch `c8y_ops` & `c8y_ops2` (Assuming the files were not existing earlier) then the watcher will be added to the directory. When some event triggers on this directory, then we need to check on which files these events were raised. If the event is raised for the files we are interested in then we can capture those events and forward them for further processing.

Reply:
Okay it's clearer now. However, this deserves a comment describing the structure and the why.

Reply:
added a comment